### DISCUSSION TITLE
correct usage of getsockname() and getpeername()

### DISCUSSION THREAD

**Discussion by faust22** [ANLAB-KAIST/KENSv3#132] [Project Q&A 1. Application Layer (Socket)] [answered]
I'm in Window+WSL2 environment and got a weird result
<img width="344" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/106755574/844f5a8f-4d90-4d62-8624-067993231d55">
<img width="341" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/106755574/a5032dd3-0428-4487-91cb-e0f314e6b33e">

I thought getsockname() gives its caller process' file descriptor bonded ip, however it makes "2.0.~.~" ip instead of "1.0.~.~" testcases.
Do KENS somehow use kinds of virtual address? or was it just my mistake?

**Comment by Pusnow** [comment]
The binary named `echo-non-kens` does not run on KENS, and thus, `getsockname()` returns WSL VM's IP addresses. This binary is only for testing on real network environments.

The another binary named `echo` is for testing on KENS environment. Please use this binary for testing.

**Reply by faust22** [reply]
<img width="284" alt="image" src="https://github.com/ANLAB-KAIST/KENSv3/assets/106755574/0a0ad8d5-7103-49e3-9a3f-640dd3ffeb60">
response and content are debugging info, which are printed to tell us what they exactly found. So, this time I executed KENS version of echo and I still confronted same problem, getsockname()(or getpeername()) still returns virtual address(2.0.49.97). I guess that in my program, server & client answer disappeared because of it (different ip_address find different contents in std::map<std::string, std::string> &___answers). is it impossible to solve it on WSL2?

**Reply by Pusnow** [reply]
Please check you are using KENS's `getsockname`/`getpeername`.

If you include `sys/socket.h` header file, your code possibly uses Linux's ones, so please do *not* include the header file.

Also, you must use the KENS's implementations in

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/src/Networking/TCP/E_TCPApplication.cpp#L40-L59

You can insert `printf`s in the code blocks and check you are using the KENS's implementations.


KENS creates and only uses a *virtual* network inside a process, so the external network environments (IP addresses, etc.) do not affect results.

**Comment by Pusnow** [comment, marked as answer]
> I'm still having trouble, getsockname(), getpeername() gives me strange ip address starting "2.0.~.~" even after you reply to my question.
> I used correct functions defined in KENS. (https://github.com/ANLAB-KAIST/KENSv3/discussions/132)
> 
> I'm trying debugging, and I found out this happens when it is on my virtual machine(VMware station).
> I want to try it on different network or different laptop but I don't have any Mac or second devices based on UNIX.
> So, can you just compile my works and execute it in your environment and tell me whether it's still in trouble or not?
> Thank you.

I found that you are using `inet_ntop()` incorrectly. The second parameter of `inet_ntop()` is `sin_addr` of `struct sockaddr_in`, not `struct sockaddr_in` itself.

```c
inet_ntop(AF_INET, &server_address, buf, BUF_SIZE-1)
```

should be

```c
inet_ntop(AF_INET, &server_address.sin_addr, buf, BUF_SIZE-1) 
```

**Reply by faust22** [reply]
very thank you!!!
